matchms/filtering/SpectrumProcessor.py

```diff
@@ -24,116 +23,106 @@ class SpectrumProcessor:
-            if filter_name not in FILTER_FUNCTION_NAMES:
-                raise ValueError(f"Unknown filter type: {filter_name} Should be known filter name or function.")
-            new_filter_func = partial(FILTER_FUNCTION_NAMES[filter_name], **filter_args)
-            new_filter_func.__name__ = FILTER_FUNCTION_NAMES[filter_name].__name__
+        filter_args = None
+        if isinstance(filter_description, (tuple, list)):
+            if len(filter_description) == 1:
+                filter_function = filter_description[0]
+            elif len(filter_description) == 2:
+                filter_function = filter_description[0]
+                filter_args = filter_description[1]
+            else:
+                raise ValueError("The filter_function should contain only two values, "
+                                 "the first should be string or callable and the second a dictionary with settings")
         else:
-            raise TypeError("filter_spec should be a string or a tuple or list")
-        check_all_parameters_given(new_filter_func)
-
+            filter_function = filter_description
+        if isinstance(filter_function, str):
+            filter_function = load_matchms_filter_from_string(filter_function)
+        self._add_filter_to_filter_order(filter_function.__name__,
+                                         filter_position=filter_position)
+        self._store_filter(filter_function, filter_args)
+
+    def _store_filter(self,
+                      new_filter_function: Callable,
+                      filter_params: Optional[Dict[str, any]]):
+        """Stores filter, removes duplicates and sorts filters"""
+        if not callable(new_filter_function):
+            raise TypeError("Expected callable filter function.")
+        new_filter_function = create_partial_function(new_filter_function, filter_params)
+        check_all_parameters_given(new_filter_function)
         # Replace filters that are already stored.
         filter_already_added = False
         for i, filter_function in enumerate(self.filters):
-            if new_filter_func.__name__ == filter_function.__name__:
+            if new_filter_function.__name__ == filter_function.__name__:
                 logger.warning("The filter %s was already in the filter list, "
                                "the last added filter parameters are used, "
-                               "check yaml file for details", new_filter_func)
-                self.filters[i] = new_filter_func
+                               "check yaml file for details", new_filter_function.__name__)
+                self.filters[i] = new_filter_function
                 filter_already_added = True
         if not filter_already_added:
-            self.filters.append(new_filter_func)
+            self.filters.append(new_filter_function)
         # Sort filters according to their order in self.filter_order
         self.filters.sort(key=lambda f: self.filter_order.index(f.__name__))
```

Comment:
This whole section is a bit convoluted and hiding what it is actually doing. Putting all the checks into functions would make it easier to read.

Reply:
Agreed. i have added a _replace_already_stored_filters function